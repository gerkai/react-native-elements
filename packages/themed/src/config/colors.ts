import {
  Colors,
  darkColors,
  lightColors,
} from '@react-native-elements/base/dist/helpers';

export { lightColors, darkColors };
export type { Colors };
