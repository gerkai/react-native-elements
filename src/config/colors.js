import { StyleSheet } from 'react-native';

export default {
  primary: '#2089dc',
  secondary: '#8F0CE8',
  grey0: '#393e42',
  grey1: '#43484d',
  grey2: '#5e6977',
  grey3: '#86939e',
  grey4: '#bdc6cf',
  grey5: '#e1e8ee',
  greyOutline: '#bbb',
  searchBg: '#303337',
  error: '#ff190c',
  disabled: 'hsl(208, 8%, 90%)',
  // Darker color if hairlineWidth is not thin enough
  divider: StyleSheet.hairlineWidth < 1 ? '#bcbbc1' : 'rgba(0, 0, 0, 0.12)',
};
