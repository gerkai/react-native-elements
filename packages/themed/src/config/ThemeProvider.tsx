import React, { useCallback, useContext } from 'react';
import deepmerge from 'deepmerge';
import { Colors, lightColors, darkColors } from './colors';
import {
  ThemeMode,
  RecursivePartial,
  Theme,
  ThemeSpacing,
  defaultSpacing,
} from './theme';
import { ComponentTheme } from './theme.component';

export type { RecursivePartial };

type ComponentFunctionProps<Components = ComponentTheme> = {
  [Key in keyof Components]?:
    | Components[Key]
    | ((props: Components[Key]) => Components[Key]);
};

export interface CreateThemeOptions
  extends ComponentFunctionProps,
    RecursivePartial<Theme> {
  lightColors?: RecursivePartial<Colors>;
  darkColors?: RecursivePartial<Colors>;
}

export interface ThemeOptions extends ComponentFunctionProps, Theme {
  colors: Colors;
}

export type UpdateTheme = (
  myNewTheme:
    | CreateThemeOptions
    | ((myTheme: CreateThemeOptions) => CreateThemeOptions)
) => void;

export type ReplaceTheme = (
  updates:
    | CreateThemeOptions
    | ((myTheme: CreateThemeOptions) => CreateThemeOptions)
) => void;

export type ThemeProps<T = {}> = {
  theme: ThemeOptions & T;
  updateTheme: UpdateTheme;
  replaceTheme: ReplaceTheme;
};

export type ThemeProviderContext<T = {}> = ThemeProps<ThemeOptions & T>;

export const ThemeContext = React.createContext<ThemeProviderContext>(
  {} as ThemeProviderContext
);

export const createTheme = (
  theme: CreateThemeOptions = {}
): CreateThemeOptions => {
  return {
    ...theme,
    ...deepmerge<CreateThemeOptions>(
      { lightColors, darkColors, spacing: defaultSpacing },
      {
        lightColors: theme.lightColors || ({} as Colors),
        darkColors: theme.darkColors || ({} as Colors),
        mode: theme.mode || 'light',
        spacing: theme.spacing || {},
      }
    ),
  };
};

const separateColors = (
  theme: CreateThemeOptions,
  themeMode?: ThemeMode
): ThemeOptions => {
  const {
    darkColors: themeDarkColors = {},
    lightColors: themeLightColors = {},
    spacing = {},
    mode = themeMode,
    ...restTheme
  } = theme;

  const themeColors = mode === 'dark' ? themeDarkColors : themeLightColors;
  return {
    colors: themeColors as Colors,
    mode,
    spacing: spacing as ThemeSpacing,
    ...restTheme,
  };
};

export const ThemeProvider: React.FC<{
  theme?: CreateThemeOptions;
  children?: React.ReactNode;
}> = ({ theme = createTheme({}), children }) => {
  const [themeState, setThemeState] = React.useState<CreateThemeOptions>(theme);

  const updateTheme: UpdateTheme = React.useCallback((updatedTheme) => {
    setThemeState((oldTheme) => {
      const newTheme =
        typeof updatedTheme === 'function'
          ? updatedTheme(oldTheme)
          : updatedTheme;
      return deepmerge({ ...oldTheme }, newTheme);
    });
  }, []);

  const replaceTheme: ReplaceTheme = React.useCallback((replacedTheme) => {
    setThemeState((oldTheme) => {
      const newTheme =
        typeof replacedTheme === 'function'
          ? replacedTheme(oldTheme)
          : replacedTheme;
      return deepmerge(createTheme({}), newTheme);
    });
  }, []);

  const ThemeContextValue = React.useMemo(
    () => ({
      theme: separateColors(themeState, themeState.mode),
      updateTheme,
      replaceTheme,
    }),
    [themeState, updateTheme, replaceTheme]
  );

  return (
    <ThemeContext.Provider value={ThemeContextValue}>
      {children}
    </ThemeContext.Provider>
  );
};

export const ThemeConsumer = ThemeContext.Consumer;

interface UseTheme {
  replaceTheme: ReplaceTheme;
  updateTheme: UpdateTheme;
  theme: {
    colors: Colors;
  } & Theme;
}

export const useTheme = (): UseTheme => {
  return useContext(ThemeContext);
};

export const useThemeMode = () => {
  const {
    updateTheme,
    theme: { mode },
  } = useTheme();

  const setMode = useCallback(
    (colorMode: ThemeMode) => {
      updateTheme({ mode: colorMode });
    },
    [updateTheme]
  );

  return { mode, setMode };
};
