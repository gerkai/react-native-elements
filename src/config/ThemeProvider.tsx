import React from 'react';
import deepmerge from 'deepmerge';
import colors from './colors';
import darkColors from './colorsDark';
import { FullTheme, Theme } from './theme';

type RecursivePartial<T> = { [P in keyof T]?: RecursivePartial<T[P]> };

export interface ThemeProps<T> {
  theme: Theme<T>;
  updateTheme: (updates: RecursivePartial<FullTheme>) => void;
  replaceTheme: (updates: RecursivePartial<FullTheme>) => void;
}

export const ThemeContext: React.Context<ThemeProps<{}>> = React.createContext({
  theme: {
    colors,
  },
} as ThemeProps<{}>);

export type ThemeProviderProps = {
  useDark?: boolean;
};

type ThemeProviderState = {
  theme: Theme;
  useDark: boolean;
};

export default class ThemeProvider extends React.Component<
  ThemeProviderProps,
  ThemeProviderState
> {
  static defaultProps = {
    theme: {},
    useDark: false,
  };
  defaultTheme: Partial<FullTheme>;

  constructor(props: { theme: Theme; useDark?: boolean }) {
    super(props);
    const defaultColors = props.useDark ? darkColors : colors;
    this.defaultTheme = deepmerge(
      {
        colors: defaultColors,
      },
      props.theme
    );
    this.state = {
      theme: this.defaultTheme,
      useDark: Boolean(props.useDark),
    };
  }

  static getDerivedStateFromProps(
    props: {
      theme: Theme;
      useDark?: boolean;
    },
    state: ThemeProviderState
  ) {
    const { useDark } = props;
    if (useDark !== state.useDark) {
      const defaultColors = useDark ? darkColors : colors;
      return {
        theme: deepmerge(
          state.theme,
          deepmerge(
            {
              colors: defaultColors,
            },
            props.theme
          )
        ),
        useDark,
      };
    }
    return null;
  }

  updateTheme = (updates: RecursivePartial<FullTheme>) => {
    this.setState(({ theme }) => ({
      theme: deepmerge(theme, updates),
    }));
  };

  replaceTheme = (theme: RecursivePartial<FullTheme>) => {
    this.setState(() => ({
      theme: deepmerge(this.defaultTheme, theme),
    }));
  };

  getTheme = () => this.state.theme;

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.state.theme,
          updateTheme: this.updateTheme,
          replaceTheme: this.replaceTheme,
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export const ThemeConsumer = ThemeContext.Consumer;
