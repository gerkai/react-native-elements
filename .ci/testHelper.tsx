import React from 'react';
import { render } from '@testing-library/react-native';
import { ThemeProvider, FullTheme } from '../src/config';

const theme = {
  colors: {
    black: 'black',
  },
};

export const renderWithTheme = (
  children: any,
  themeProp: Partial<FullTheme> = theme
) => {
  return render(<ThemeProvider theme={themeProp}>{children}</ThemeProvider>);
};
