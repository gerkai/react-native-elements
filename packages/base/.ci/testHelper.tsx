import React, { JSXElementConstructor } from 'react';
import { View } from 'react-native';
import {
  render,
  RenderOptions,
  fireEvent,
  act,
} from '@testing-library/react-native';
import { ThemeProvider, FullTheme, colors } from '../src/config';
import deepmerge from 'deepmerge';

export { fireEvent, act };

// for getting findByType e.g. wrapper.findByType(Icon) see implementation in Avatar Component
export const renderWithWrapper = (
  children: React.ReactElement<any, string | JSXElementConstructor<any>>,
  wrapperTestID?: string,
  themeProp: Partial<FullTheme> = {},
  renderOptions?: RenderOptions
) => {
  const options: RenderOptions = {
    ...(!wrapperTestID && {
      wrapper: (props) => <View {...props} testID="wrapper" />,
    }),
    ...renderOptions,
  };
  const renderApi = render(
    themeProp ? (
      <ThemeProvider theme={deepmerge({ colors }, themeProp)}>
        {children}
      </ThemeProvider>
    ) : (
      children
    ),
    options
  );
  const wrapper = renderApi.queryByTestId(wrapperTestID || 'wrapper');
  return { wrapper, ...renderApi };
};
