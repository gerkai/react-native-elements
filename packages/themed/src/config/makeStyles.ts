import { useContext, useMemo } from 'react';
import { StyleSheet } from 'react-native';
import { Colors } from './colors';
import { FullTheme, Theme, ThemeMode } from './theme';
import { ReplaceTheme, ThemeContext, UpdateTheme } from './ThemeProvider';

interface useThemeI {
  replaceTheme: ReplaceTheme;
  updateTheme: UpdateTheme;
  theme: {
    colors: Colors;
    mode: ThemeMode;
  } & Theme;
}

export const useTheme = (): useThemeI => {
  return useContext(ThemeContext);
};

export const useThemeMode = () => {
  return useContext(ThemeContext).theme.mode;
};

export const makeStyles =
  <T extends StyleSheet.NamedStyles<T> | StyleSheet.NamedStyles<any>, V>(
    styles: T | ((theme: Partial<FullTheme>, props: V) => T)
  ) =>
  (props: V = {} as any): T => {
    const { theme } = useTheme();

    return useMemo(() => {
      const css =
        typeof styles === 'function'
          ? styles({ colors: theme.colors, mode: theme.mode }, props)
          : styles;
      return StyleSheet.create(css);
    }, [props, theme.colors, theme.mode]);
  };
