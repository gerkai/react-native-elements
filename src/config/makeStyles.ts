import { useContext } from 'react';
import { StyleSheet } from 'react-native';
import { FullTheme } from './theme';
import { ThemeContext } from './ThemeProvider';

export const useTheme = () => {
  return useContext(ThemeContext);
};

export const makeStyles = <
  T extends StyleSheet.NamedStyles<T> | StyleSheet.NamedStyles<any>,
  V
>(
  styles: T | ((theme: Partial<FullTheme>, props: V) => T)
) => (props: V = {} as any): T => {
  const { theme } = useTheme();
  const css = typeof styles === 'function' ? styles(theme, props) : styles;

  return StyleSheet.create(css);
};
