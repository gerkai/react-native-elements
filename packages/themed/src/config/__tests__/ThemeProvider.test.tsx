import React from 'react';
import { Text, Button } from '../..';
import { colors } from '../colors';
import { renderWithWrapper } from '../../../.ci/testHelper';
import { fireEvent, render } from '@testing-library/react-native';
import { useTheme } from '../makeStyles';
import { ThemeProvider, createTheme, ThemeConsumer } from '../ThemeProvider';
import { View } from 'react-native';

describe('ThemeProvider', () => {
  it('render ThemeProvider', () => {
    const { toJSON } = renderWithWrapper(<Text />);
    expect(toJSON).toMatchSnapshot();
  });

  it('should update and replace theme', () => {
    const TestComp = (): JSX.Element => {
      const { theme, replaceTheme, updateTheme } = useTheme();
      return (
        <>
          <Button
            testID="updateTheme"
            onPress={() => {
              updateTheme({
                colors: {
                  primary: 'red',
                },
              });
            }}
          />
          <Button
            testID="replaceThemeButton"
            onPress={() => {
              replaceTheme({
                colors: {
                  primary: 'purple',
                },
              });
            }}
          />
          <Text testID="themeChild" children={theme.colors.primary} />
        </>
      );
    };

    const { queryByTestId } = render(
      <ThemeProvider>
        <TestComp />
      </ThemeProvider>
    );
    const updateButton = queryByTestId('updateTheme');
    const replaceButton = queryByTestId('replaceThemeButton');
    const textTheme = queryByTestId('themeChild');
    expect(textTheme.props.children).toEqual(colors.primary);

    fireEvent.press(updateButton);
    expect(textTheme.props.children).toEqual('red');

    fireEvent.press(replaceButton);
    expect(textTheme.props.children).toBe('purple');
  });

  it('should use default theme', () => {
    const { queryByTestId } = renderWithWrapper(
      <ThemeProvider>
        <ThemeConsumer>
          {({ theme }) => (
            <View testID="viewComp">{JSON.stringify(theme)}</View>
          )}
        </ThemeConsumer>
      </ThemeProvider>
    );
    const instance = queryByTestId('viewComp');
    expect(JSON.parse(instance.props.children)).toMatchObject({
      colors,
    });
  });

  it('should retain custom theme when switching between light / dark mode', () => {
    const customTheme = createTheme({
      colors: {
        primary: 'white',
      },
      darkColors: {
        primary: 'black',
      },
      mode: 'light',
      Text: { accessibilityLabel: 'theme-test' },
    });
    const TestComp = (): JSX.Element => {
      const { theme, updateTheme } = useTheme();

      return (
        <>
          <Button
            testID="updateTheme"
            onPress={() =>
              updateTheme((myTheme) => ({
                mode: myTheme.mode === 'dark' ? 'light' : 'dark',
              }))
            }
          />
          <Text testID="themeChild" children={String(theme.mode)} />
        </>
      );
    };
    const { queryByTestId } = render(
      <ThemeProvider theme={customTheme}>
        <TestComp />
      </ThemeProvider>
    );
    const updateButton = queryByTestId('updateTheme');
    const textTheme = queryByTestId('themeChild');

    expect(textTheme.props.accessibilityLabel).toEqual('theme-test');

    // set to dark mode
    fireEvent.press(updateButton);
    expect(textTheme.props.accessibilityLabel).toEqual('theme-test');
    expect(textTheme.props.children).toEqual('dark');

    // set to light mode
    fireEvent.press(updateButton);
    expect(textTheme.props.accessibilityLabel).toEqual('theme-test');
    expect(textTheme.props.children).toEqual('light');
  });
});
