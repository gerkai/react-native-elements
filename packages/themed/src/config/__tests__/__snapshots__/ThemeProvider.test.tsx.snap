// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeProvider render ThemeProvider 1`] = `
Array [
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 24,
        },
        undefined,
        false,
      ]
    }
    testID="RNE_BUTTON_WRAPPER"
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="xl"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 24,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
            "paddingHorizontal": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Test
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 12,
        },
        undefined,
        false,
      ]
    }
    testID="RNE_BUTTON_WRAPPER"
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="lg"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 12,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
            "paddingHorizontal": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Test
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 4,
        },
        undefined,
        false,
      ]
    }
    testID="RNE_BUTTON_WRAPPER"
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="sm"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 4,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
            "paddingHorizontal": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Test
        </Text>
      </View>
    </View>
  </View>,
  <View
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 8,
        },
        undefined,
        false,
      ]
    }
    testID="RNE_BUTTON_WRAPPER"
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="md"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#2089dc",
            "borderColor": "#2089dc",
            "borderRadius": 8,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
            "paddingHorizontal": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "white",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Test
        </Text>
      </View>
    </View>
  </View>,
]
`;
